### PR TITLE
Write Ballerina Central Stats to Sheets

### PR DESCRIPTION
**Pull request by NirashaSewwandi**
## Purpose
- https://github.com/ballerina-platform/ballerina-central/issues/95

### CODE REVIEW

File: centralStats/main.bal [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+# + return - Rows to be inserted to the excel sheet
+#
+public function writeDataToSheet(HttpResponse response, sheets:Client spreadsheetClient, string spreadsheetID, string sheetName) returns error? {
+
+    string[] data = [];
+    string[][] excelRows = [];
+    Tables tableResult = <Tables>response.tables[0];
+
+    foreach string[] row in tableResult.rows {
+        data = row;
```
Comment (xlight05) [inline centralStats/main.bal:159]: do we need this?

File: centralStats/main.bal [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+    sheets:A1Range a1Range = {
+        sheetName: sheetName
+    };
+    foreach string[] values in excelRows {
+        _ = check spreadsheetClient->appendValue(spreadsheetID, values, a1Range);
+    }
+
+    return ();
```
Comment (xlight05) [inline centralStats/main.bal:171]: Not required to return nil

**Review by xlight05 — CHANGES_REQUESTED** [review]

File: .github/workflows/ballerina-central-stats.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+run-name: Testing out GitHub Actions Now
+on: [push]
+# on:
+#   workflow_dispatch:
+#   schedule:
+#     - cron: '30 18 * * *'   # 00:00 in LK time (GMT+5:30)
+
+
+jobs:
+  nballerina-build:
```
Comment (xlight05) [inline .github/workflows/ballerina-central-stats.yml:11]: change names to valid ones :D

File: centralStats/build [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
```
[the remark this answers is not in the document]
Comment (NirashaSewwandi) [reply centralStats/build]: No

File: centralStats/Ballerina.toml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+[package]
+org = "sns"
+name = "dashboardExtractor"
```
[the remark this answers is not in the document]
Comment (NirashaSewwandi) [reply centralStats/Ballerina.toml]: Fixed https://github.com/ballerina-platform/ballerina-release/pull/2576/commits/dfc15dbb74e88ba2c5a0393162b2316a9ce38169

File: centralStats/main.bal [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+| summarize PullCount = tostring(count()) by tostring(customDimensions["name"]) ,  org = tostring(customDimensions["organization"])
+| sort by PullCount desc;
+x ;`;
+
+string SPREADSHEET_CLIENT_ID = os:getEnv("SPREADSHEET_CLIENT_ID");
+string SPREADSHEET_CLIENT_SECRET = os:getEnv("SPREADSHEET_CLIENT_SECRET");
+string SPREADSHEET_REFRESH_TOKEN = os:getEnv("SPREADSHEET_REFRESH_TOKEN");
+string SPREADSHEET_ID = os:getEnv("SPREADSHEET_ID");
+string applicationID = os:getEnv("APPLICATION_ID");
+string apiKey = os:getEnv("API_KEY");
```
[the remark this answers is not in the document]
Comment (NirashaSewwandi) [reply centralStats/main.bal]: Fixed https://github.com/ballerina-platform/ballerina-release/pull/2576/commits/dfc15dbb74e88ba2c5a0393162b2316a9ce38169

File: centralStats/main.bal (hunk 2 above)
Comment (NirashaSewwandi) [reply centralStats/main.bal]: Fixed https://github.com/ballerina-platform/ballerina-release/pull/2576/commits/dfc15dbb74e88ba2c5a0393162b2316a9ce38169

File: .github/workflows/ballerina-central-stats.yml (hunk 3 above)
Comment (NirashaSewwandi) [reply .github/workflows/ballerina-central-stats.yml]: Fixed :) https://github.com/ballerina-platform/ballerina-release/pull/2576/commits/dfc15dbb74e88ba2c5a0393162b2316a9ce38169

File: .github/workflows/ballerina-central-stats.yml [hunk 7]
```diff
@@ -0,0 +1,30 @@
+name: Dashboard Extractor
+run-name: Testing out GitHub Actions
+on: [push]
```
Comment (anuruddhal) [inline .github/workflows/ballerina-central-stats.yml:3]: Is this trigger correct?

File: central_stats/main.bal [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+            refreshToken: SPREADSHEET_REFRESH_TOKEN
+        }
+    };
+
+    sheets:Client spreadsheetClient = check new (spreadsheetConfig);
+
+    // pull count - ballerina/ballerinax
+    string encodedQuery = check url:encode(queryPullCountOfBallerinaBallerinax, "UTF-8");
+    HttpResponse response = check http->/query.get({
+            "x-api-key": apiKey
```
Comment (anuruddhal) [inline central_stats/main.bal:112]: Move `x-api-key` to a constant.

**Review by anuruddhal — CHANGES_REQUESTED** [review]

File: central_stats/main.bal (hunk 8 above)
Comment (NirashaSewwandi) [reply central_stats/main.bal]: Fixed in https://github.com/ballerina-platform/ballerina-release/pull/2576/commits/b351f15e1d0a2d11b1ce8f7b32e100830175825d

File: .github/workflows/ballerina-central-stats.yml (hunk 7 above)
Comment (NirashaSewwandi) [reply .github/workflows/ballerina-central-stats.yml]: Fixed in https://github.com/ballerina-platform/ballerina-release/pull/2576/commits/b351f15e1d0a2d11b1ce8f7b32e100830175825d

File: centralStats/main.bal (hunk 1 above)
Comment (NirashaSewwandi) [reply centralStats/main.bal]: No. Removed but for loop is there for row by row writing

File: centralStats/main.bal [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+            refreshToken: SPREADSHEET_REFRESH_TOKEN
+        }
+    };
+
+    sheets:Client spreadsheetClient = check new (spreadsheetConfig);
+
+    // pull count - ballerina/ballerinax
+    string encodedQuery = check url:encode(queryPullCountOfBallerinaBallerinax, "UTF-8");
+    string path = string `/query?query=${encodedQuery}`;
+    HttpResponse response = check http->get(path, headers);
```
[the remark this answers is not in the document]
Comment (NirashaSewwandi) [reply centralStats/main.bal]: Fixed

**Review by anuruddhal — APPROVED** [review]

**Review by xlight05 — APPROVED** [review]

**Pull request merged**